File: erpnext/accounts/report/wbs_drilldown/wbs_drilldown.js
// Copyright (c) 2023, 8848 Digital LLP and contributors
// For license information, please see license.txt
/* eslint-disable */


frappe.query_reports["WBS Drilldown"] = {
	"filters": [

		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd":1

		},

		

		// {
		// 	"fieldname":"project",
		// 	"label": __("Project"),
		// 	"fieldtype": "Link",
		// 	"options":'Project',
		// 	"reqd":1			
		// },

		{
			"fieldname":"project",
			"label": __("Project"),
			"fieldtype": "MultiSelectList",
			"options": "Project",
			get_data: function(txt) {
				return frappe.db.get_link_options('Project', txt, {
					is_wbs:1,
					company:frappe.query_report.get_filter_value("company")

				});
			}
		},

		{
			"fieldname":"wbs",
			"label": __("WBS"),
			"fieldtype": "Link",
			"options":'Work Breakdown Structure'
			


			
		},

		{
			"fieldname":"show_group_totals",
			"label": __("Show Group Totals"),
			"fieldtype": "Check",
			"default":1
		}

		
		

	],

	"open_budget_entry":function(data){
		if (!data.wbs) return;
		let project = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'project'; });

		var only_overall_amounts = 0
		if(data.only_overall_amounts = 1){
			only_overall_amounts = 1
		}
		else{
			only_overall_amounts = 0
		}
		frappe.route_options = {
			"wbs": data.wbs,
			"company": frappe.query_report.get_filter_value('company'),
			"project": (project && project.length > 0) ? project[0].$input.val() : ""
		};

		let report = "Budget Entry";

		

		frappe.set_route("query-report", report);
	},

	"open_budget_entry2":function(data){
		if (!data.wbs) return;
		let project = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'project'; });
		let group_totals = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'show_group_totals'; });
		if(!project){
			project = data.project
		}

		// var html = `<p>This is a lock symbol: &#128274;</p>`
		// console.log(html)
		// var all_wbs = []
		// if(group_totals[0].value == 1){
		// 	frappe.call({
		// 		method:"project_system.project_system.report.wbs_drilldown.wbs_drilldown.get_all_wbs",
		// 		args:{
		// 			"wbs":data.wbs
		// 		},
		// 		callback:function(r){
		// 			if(r.message){
		// 				all_wbs = r.message
		// 			}
		// 			console.log("------------------------wbs34567",all_wbs)
		// 		}
		// 	})
		// }
		// else{
		// 	all_wbs = data.wbs
		// }
	
		frappe.route_options = {
			"wbs": data.wbs,
			"company": frappe.query_report.get_filter_value('company'),
			"only_overall_amounts":1,
			"show_group_totals":group_totals[0].value,
			"project": (project && project.length > 0) ? project[0].$input.val() : ""
		};

		let report = "Budget Entry";

		

		frappe.set_route("query-report", report);
	},


	"open_committed_line_items":function(data){
		if (!data.wbs) return;
		let project = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'project'; });
		let group_totals = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'show_group_totals'; });

		
		frappe.route_options = {
			"wbs": data.wbs,
			"show_group_totals":group_totals[0].value,
			"project": (project && project.length > 0) ? project[0].$input.val() : ""
		};

		let report = "Commited Line Item";

		

		frappe.set_route("query-report", report);
	},

	"open_actual_line_items":function(data){
		if (!data.wbs) return;
		let project = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'project'; });
		let group_totals = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'show_group_totals'; });

		
		frappe.route_options = {
			"wbs": data.wbs,
			"show_group_totals":group_totals[0].value,
			"project": (project && project.length > 0) ? project[0].$input.val() : ""
		};

		let report = "Actual Line Items";

		

		frappe.set_route("query-report", report);
	},

	


	"formatter": function(value, row, column, data, default_formatter) {
		const cname = ["wbs","available_budget"]
		if (data && column.fieldname == "wbs") {
			data.only_overall_amounts = 0
			value = data.wbs || value;

			column.link_onclick =
			"frappe.query_reports['WBS Drilldown'].open_budget_entry(" + JSON.stringify(data) + ")";
			column.is_tree = true;
		}

		else if (data && column.fieldname == "overall_budget") {
			
			value = value;
			console.log("----------------------------data2333",data)
			data.only_overall_amounts = 1
			column.link_onclick =
			"frappe.query_reports['WBS Drilldown'].open_budget_entry2(" + JSON.stringify(data) + ")";
			column.is_tree = true;
		}

		else if (data && column.fieldname == "committed_overall_budget") {
			
			value = value;
			column.link_onclick =
			"frappe.query_reports['WBS Drilldown'].open_committed_line_items(" + JSON.stringify(data) + ")";
			column.is_tree = true;
		}

		else if (data && column.fieldname == "actual_overall_budget") {
			
			value = value;
			column.link_onclick =
			"frappe.query_reports['WBS Drilldown'].open_actual_line_items(" + JSON.stringify(data) + ")";
			column.is_tree = true;
		}

		if (data && data.account && column.apply_currency_formatter) {
			data.currency = erpnext.get_currency(column.company_name);
		}

		value = default_formatter(value, row, column, data);
		if (data.created_from_project) {
			value = $(`<span>${value}</span>`);

			var $value = $(value).css("font-weight", "bold");

			value = $value.wrap("<p></p>").parent().html();
		}
		return value;
	},

	
};

File: erpnext/accounts/report/wbs_drilldown/wbs_drilldown.py
# Copyright (c) 2023, 8848 Digital LLP and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from erpnext.accounts.report.wbs_drilldown.wbs_drilldown_methods import *
from frappe.query_builder.functions import Coalesce, Sum
# frappe-bench-postgres/apps/erpnext/erpnext/accounts/report/wbs_drilldown/wbs_drilldown.py

def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data

def get_columns(filters):
    return [
        {
            "label": _("WBS Element"),
            "fieldname": "wbs",
            "fieldtype": "Link",
            "options": "Work Breakdown Structure",
            "width": 200
        },
        {
            "label": _("WBS Name"),
            "fieldname": "wbs_name",
            "fieldtype": "Data"
        },
        {
            "label": _("WBS Level"),
            "fieldname": "wbs_level",
            "fieldtype": "Data"
        },
        {
            "label": _("Created From Project"),
            "fieldname": "created_from_project",
            "fieldtype": "Check",
            "hidden": 1
        },
        {
            "label": _("Project"),
            "fieldname": "project",
            "fieldtype": "Link",
            "options": "Project",
            "width": 200
        },
        {
            "label": _("Overall Budget"),
            "fieldname": "overall_budget",
            "fieldtype": "Link",
            "options": "Work Breakdown Structure"
        },
        {
            "label": _("Committed Overall Budget"),
            "fieldname": "committed_overall_budget",
            "fieldtype": "Link",
            "options": "Work Breakdown Structure"
        },
        {
            "label": _("Actual Overall Budget"),
            "fieldname": "actual_overall_budget",
            "fieldtype": "Link",
            "options": "Work Breakdown Structure"
        },
        {
            "label": _("Assigned Overall Budget"),
            "fieldname": "assigned_overall_budget",
            "fieldtype": "Float",
            "precision": 2
        },
        {
            "label": _("Available Budget"),
            "fieldname": "available_budget",
            "fieldtype": "Float",
            "precision": 2
        }
    ]

def get_conditions(filters):
    wbs = frappe.qb.DocType("Work Breakdown Structure")
    conditions = []

    if filters.get("wbs"):
        conditions.append(wbs.name == filters.get("wbs"))

    projects = filters.get("project")
    if projects:
        if len(projects) > 1:
            conditions.append(wbs.project.isin(projects))
        else:
            conditions.append(wbs.project == projects[0])

    return conditions

def get_totals(wbs_id):
    root_wbs = frappe.get_doc("Work Breakdown Structure", wbs_id)
    totals = {
        "og_bgt": 0.0,
        "ov_bgt": 0.0,
        "cov_bgt": 0.0,
        "acov_bgt": 0.0,
        "asov_bgt": 0.0,
        "avl_bgt": 0.0
    }
    child_wbs = frappe.db.get_all("Work Breakdown Structure", {
        "lft": [">", root_wbs.get("lft")],
        "rgt": ["<", root_wbs.get("rgt")]
    }, ['name'])

    if child_wbs:
        for j in child_wbs:
            BE = frappe.qb.DocType("Budget Entry")
            query = (
                frappe.qb.from_(BE)
                .select(
                    Coalesce(Sum(BE.overall_credit - BE.overall_debit), 0.0).as_('overall_budget'),
                    Coalesce(Sum(BE.committed_overall_credit - BE.committed_overall_debit), 0.0).as_('committed_overall_budget'),
                    Coalesce(Sum(BE.actual_overall_credit - BE.actual_overall_debit), 0.0).as_('actual_overall_budget')
                )
                .where(
                    (BE.wbs == j.get("name")) &
                    ~(BE.voucher_type.isin(["Supplementary Budget", "Budget Decrease", "Budget Transfer"]))
                )
            )
            all_budget_entries = query.run(as_dict=True)

            if all_budget_entries:
                overall_budget = all_budget_entries[0].get("overall_budget")
                committed_overall_budget = all_budget_entries[0].get("committed_overall_budget")
                actual_overall_budget = all_budget_entries[0].get("actual_overall_budget")
                assigned_overall_budget = committed_overall_budget + actual_overall_budget
                available_budget = overall_budget - assigned_overall_budget

                totals["ov_bgt"] += overall_budget or 0.0
                totals["cov_bgt"] += committed_overall_budget or 0.0
                totals["acov_bgt"] += actual_overall_budget or 0.0
                totals["asov_bgt"] += assigned_overall_budget or 0.0
                totals["avl_bgt"] += available_budget or 0.0

    return totals

def get_group_wbs(filters):
    mapped_wbs = frappe._dict()
    conditions = get_conditions(filters)
    wbs = frappe.qb.DocType('Work Breakdown Structure')

    query = (
        frappe.qb.from_(wbs)
        .select(
            wbs.name, wbs.lft, wbs.rgt, wbs.parent_work_breakdown_structure.as_('parent'), 
            wbs.project, wbs.wbs_name, wbs.level, wbs.overall_budget, 
            wbs.committed_overall_budget, wbs.actual_overall_budget, 
            wbs.created_from_project, wbs.assigned_overall_budget, 
            wbs.available_budget
        )
        .where(wbs.company == filters.get("company"))
        .orderby(wbs.lft)
    )

    for condition in conditions:
        query = query.where(condition)

    all_wbs = query.run(as_dict=True)

    for d in all_wbs:
        if not filters.get("wbs"):
            if d.parent:
                mapped_wbs[d.name] = mapped_wbs.get(d.parent, 0) + 1
            else:
                mapped_wbs[d.name] = 0
        else:
            mapped_wbs[d.name] = 0

    return all_wbs, mapped_wbs

def get_data(filters):
    out = []
    applied_filters = {}

    if filters.get("project"):
        applied_filters["project"] = filters.get("project")

    if filters.get("wbs"):
        applied_filters["name"] = filters.get("wbs")

    all_wbs, mapped_wbs = get_group_wbs(filters)

    for d in all_wbs:
        row = {
            "wbs": d.name,
            'wbs_name': d.wbs_name,
            'wbs_level': d.level,
            'created_from_project': d.created_from_project,
            'project': d.project,
            'overall_budget': d.overall_budget,
            "committed_overall_budget": d.committed_overall_budget,
            "actual_overall_budget": d.actual_overall_budget,
            "assigned_overall_budget": d.assigned_overall_budget,
            "available_budget": d.available_budget,
            "indent": mapped_wbs.get(d.name),
            "parent": d.parent
        }
        out.append(row)

    if out and filters.get("show_group_totals") and not filters.get("wbs"):
        for i in out:
            if i.get("wbs") and not i.get("created_from_project"):
                totals = get_totals(i.get("wbs"))
                if totals:
                    i.update({
                        "overall_budget": i.get("overall_budget") + totals["ov_bgt"],
                        "committed_overall_budget": i.get("committed_overall_budget") + totals["cov_bgt"],
                        "actual_overall_budget": i.get('actual_overall_budget') + totals["acov_bgt"],
                        "assigned_overall_budget": i.get('assigned_overall_budget') + totals["asov_bgt"],
                        "available_budget": i.get('available_budget') + totals["avl_bgt"]
                    })

    if out:
        for i in out:
            i.update({
                "overall_budget": custom_fmt_money(i.get("overall_budget"), currency="INR"),
                "committed_overall_budget": custom_fmt_money(i.get("committed_overall_budget"), currency="INR"),
                "actual_overall_budget": custom_fmt_money(i.get('actual_overall_budget'), currency="INR"),
                "assigned_overall_budget": i.get('assigned_overall_budget'),
                "available_budget": i.get('available_budget')
            })

    return out